sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuSubPartitionHashJoin.scala

```diff
@@ -0,0 +1,649 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.sql.rapids.execution
+
+import scala.collection.mutable.ArrayBuffer
+
+import ai.rapids.cudf.{HashType, Table}
+import com.nvidia.spark.rapids.{Arm, GpuBoundReference, GpuColumnVector, GpuExpression, GpuMetric, GpuShuffledHashJoinExec, SerializedTableColumn, SpillableColumnarBatch, SpillCallback, SpillPriorities, TaskAutoCloseableResource}
+import com.nvidia.spark.rapids.RapidsPluginImplicits._
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.catalyst.plans.InnerLike
+import org.apache.spark.sql.types.StructType
+import org.apache.spark.sql.vectorized.ColumnarBatch
+
+/**
+ * Iterator that can tell if the data size of all the batches from current position
+ * is larger than the specified `targetBatchSize`, and whether the batches are
+ * serialized.
+ */
+class BatchTypeSizeAwareIterator(
+    inputIter: Iterator[ColumnarBatch],
+    targetBatchSize: Long,
+    intputBatchSize: GpuMetric) extends Iterator[ColumnarBatch]
+  with Arm with TaskAutoCloseableResource {
+
+  assert(targetBatchSize > 0,
+    s"Target batch size should not be negative, but got $targetBatchSize")
+
+  private val readBatchesQueue = ArrayBuffer.empty[ColumnarBatch]
```

Comment:
Yeah. I did some consideration on this point, and this iterator actually ends up with loading data of at most twice of the target size, that is 4GB by default, to detect the size overflow and batch type.

Anyway, it will be better to make the read batches spillable. And add it as a follow-up in  https://github.com/NVIDIA/spark-rapids/issues/7832